### ISSUE TITLE
PSR-PER 2.0.0 has been released

### ISSUE THREAD

**Issue by jrfnl** [PHPCSStandards/PHP_CodeSniffer#29]
Repost from https://github.com/squizlabs/PHP_CodeSniffer/issues/3793:

> PSR-PER is an evolving standard which builds onto PSR-12.
> 
> The first "release" (= updated standards publication) since the fork of PSR-12 has just been published.
> 
> I'm opening this ticket to allow for a discussion on if and if so, how to handle these evolving standards in PHPCS.
> 
> Refs:
> * https://github.com/php-fig/per-coding-style/releases/tag/2.0.0
> * https://github.com/php-fig/per-coding-style/compare/1.0.0...2.0.0
> * https://github.com/php-fig/per-coding-style/blob/2.0.0/spec.md
> * https://github.com/php-fig/per-coding-style/blob/2.0.0/meta.md
> 

---

@kenguest asked:

> Is there anything that can be done to help nudge this along?
> 
> Also, can I just note that the PER for Coding Standards shouldn't be referred to as a PSR - it is PER CS 2.0 :-)

---

I replied:

> @kenguest Well, aside from the decision which still needs to be taken about this, it might be good to create an action list of the differences between PSR-12 and PSR-PER ?
> 
> For each difference, it will need to be determined whether there is an existing sniff which could handle this (possibly with a `public` property setting) or whether a new sniff is needed.
> 
> Another thing to have a good think about would be how to handle versioning, as without some form of versioning, a PSR-PER ruleset would be a moving target which could randomly start failing CI builds due to new rules being introduced (and being sniffed for).

---

@kenguest replied:

> @jrfnl I'm working on a list of changes between the two standards, then an action list.
> PER-CS is at version 2 at the moment, and there will be distinct version numbers so it should just be a case of having one set of sniffs called PERCS200 (for v2.0.0) and so on.

---

My current line of thinking on this is as follows:
* I agree with @kenguest that each PER version should have its own ruleset, like `PER-200`, to prevent it from becoming a moving target failing builds at random.
    Each ruleset can include the previous version and build on top of that.
* For those repos who always want to use the latest version, there should also be a `PERCS-latest`, which would include the ruleset for the current version and would need to be updated each time a new version is added.
* As there could be a lot of PER versions, adding these rulesets to the main PHPCS repo may be a bit unwieldy, so I'm thinking a separate repo in this organisation called `PSR-PER`.
* Sniffs needed in the PER rulesets should still be pulled to the main PHPCS repo. The `PSR-PER` repo would only contain rulesets, not sniffs.

**Comment by jazzsequence** [comment]
Semantic point but I would second @kenguest's suggestion of naming and referring to the repo and standard as `PER-CS` (or a variation thereof) rather than `PSR-PER`.  **PSR** specifically refers to [PHP Standards Recommendations](https://www.php-fig.org/psr/) whereas **PER** refers to [PHP Evolving Recommendations](https://www.php-fig.org/psr/); combining the two feels like mixing separate interests (even as PER is an extension of PSR-12).

**Comment by kenguest** [comment]
Gentle reminder, @jrfnl, that there is a migration document detailing the differences between PSR-12 and PER-CS v2 at https://github.com/php-fig/per-coding-style/blob/master/migration-2.0.md should it need to be consulted for supporting PER-CS in PHP CodeSniffer.

This link hadn't been copied over from the squizlabs/phpcs ticket that I filed over a year ago.

**Comment by jrfnl** [comment]
> This link hadn't been copied over from the squizlabs/phpcs ticket that I filed over a year ago.

@kenguest Yes, that comment was posted in the gap between when I opened this issue and the official take-over announcement, which is why it wasn't copied over.

---

I'd actually been thinking off this issue not too long ago.

If it wasn't clear from the above, my preference would be for this new repo to be run by a team of ~two maintainers, with me available in the background for consultation when needed.

The "vacancy" description for these roles would be as follows:

### Tasks:

#### Initial set up of the PER-CS repo

* Set up the initial PER 1.0.0, PER 2.0.0, PER-latest rulesets.
* Find sniffs for those rules which were added/changed in PER 2.0.0 compared to PER 1.0.0 (PSR-12).
    The migration guide @kenguest linked to should be used as a basis for this.
* Identify if there are rules which cannot be covered via the currently available sniffs and if so, open a feature request in this repo for each uncovered rule with a clear description of the specifications for the sniff which would need to be written.
    Where needed, ask for clarification from the PER team if there is any ambiguity as a sniff needs exact specifications to follow.
* Set up CI/QA checks for the rulesets.
* Register the package with Packagist.
* Tag the initial release.

#### Continued maintenance of the PER-CS repo

* Follow/watch this repo, in particular the PER related feature request tickets and new releases, to know when new sniffs are available which could fill in the gaps.
* When "missing" sniffs become available, make the appropriate updates to the repo and tag a new release.
* Follow/watch the [php-fig/per-coding-style](https://github.com/php-fig/per-coding-style) repo.
    Carefully scrutinize change proposals for ambiguity and ask for clarification where needed.
* When a new PER release gets tagged, follow the recipe above (add a new ruleset, add the new ruleset to CI in all the appropriate places, find sniffs to add to it to cover the difference with the previous version, identify missing sniffs and open feature requests for those).

#### Optional additional tasks (time and skills permitting)

* Verify that the sniffs used in the PER rulesets all have documentation and contribute to adding documentation if any is missing (also see #148)
* Potentially volunteer to be part of the PER team ?
* Potentially address the opened feature requests via PRs.
* Potentially address gaps in the PER rules, like when new syntaxes get introduced in PHP, and open issues in the PER repo with standards suggestions/the suggestion to add rules for the new PHP syntaxes.


### Time investment

For the initial setup, I'd estimate approximately a week's work.
For the continued maintenance, I'd estimate 1 - 2 hours a month.
For the optional tasks... I don't dare to put an estimate on that.


### Planning

The call for volunteers for the above is open.

If anyone is interested, but wants help getting started, I'm happy to plan a few pairing sessions via videocall to help with the initial setup.

If there are several people who are potentially interested, but not yet sure if they are up for it, I'd be open to a mob programming session to help people get more familiar with the moving parts/answer any questions/explain whatever needs explaining.


### Call to action

Anyone who is interested in getting involved with this, please, please leave a comment on this ticket.

**Comment by stankarp** [comment]
@jrfnl Any update on this issue? Have you managed to setup a new PER-CS repo or find someone who willing to volunteer?

**Comment by jrfnl** [comment]
@stankarp If there had been volunteers, you would have seen their comments in this ticket (or a comment by me that people had reached out privately), so no progress so far, unfortunately.

My offer to help get this off the ground with some videocalls to get the maintainers started is still open.

**Comment by rikvdh** [comment]
@jrfnl I'm considering supporting this. I do not have loads of spare time, but I would love to work on this. According to your 'to do' list above, my first question would be. Is this desirable in a separate repo, or integrated into this repo?

**Comment by jrfnl** [comment]
> @jrfnl I'm considering supporting this. I do not have loads of spare time, but I would love to work on this. According to your 'to do' list above, my first question would be. Is this desirable in a separate repo, or integrated into this repo?

That would be great!

It will be in a separate repo, but, as far as I'm concerned, the repo will be in the PHPCSStandards organisation.

The reasons for having a separate repo is quite straight-forward:
* I don't want an ever growing list of PER-x.x.x standards in this repo. As it is a living standards there could easily be a dozen PER standards (one for each version), which is unwieldy and undesirable in this repo.
* The skill level needed to maintain the PER Standards package is different from the skill level needed to maintain PHPCS itself. Having it in a separate repo will allow me to give potential maintainers commit/maintain rights to that repo, without them having those rights on the PHPCS repo.
* It will also allow for a more hands-off approach from me if other people take ownership of that repo.
* Aside from all that, it would also allow the package to - if necessary - use other external dependencies, aside from those used by PHPCS. This is currently not an option for PHPCS itself. I can imagine, for instance, that a dependency on PHPCSExtra could be useful depending on the sniff needs of the project.

Does that answer your question ?

**Comment by rikvdh** [comment]
Fair enough.

PER 1.0.0 is identical to PSR-12, so this will be an "easy fix".

What is a good approach to get started? Start a repo under my own name and then migrate later to the PHPCSStandards organisation?

**Comment by jrfnl** [comment]
> Fair enough.
> 
> PER 1.0.0 is identical to PSR-12, so this will be an "easy fix".

Correct, PER 2.0.0 is the bit which needs work (and setting up the repo and CI etc).

> What is a good approach to get started? Start a repo under my own name and then migrate later to the PHPCSStandards organisation?

_grin_ I actually already created a private repo in this organisation for it. Let me see if I can invite you to that.

Would you like to set up a call to talk things through ? If so, probably best to sort that out off-GH. You can DM me on [Mastodon](https://phpc.social/@jrf_nl), [X](https://x.com/jrf_nl) or - seeing your location - on the [PHPNL Slack](https://phpnl.nl/).

**Comment by rikvdh** [comment]
I'll get in touch tomorrow to schedule something. :-)

**Comment by VSlokker** [comment]
I don't have too much spare time but willing to help out on some of the work.
I also don't have too much experience contributing to open source but i have to start somewhere :man_shrugging: 
Let me know if there is anything i can do.

**Comment by aurimasbutkus** [comment]
I am also interested in helping with at least some of the work to get this up and running

**Comment by jrfnl** [comment]
@VSlokker @aurimasbutkus Thank you both for your offer to help with this.

To get you started, would it help to schedule a call ? Could you both DM me on [Mastodon](https://phpc.social/@jrf_nl) or [X](https://x.com/jrf_nl) to schedule something ?

If there is anyone else watching this ticket who would like to join the call, please do the same.

As mentioned above, the standard will be developed in a separate repo. @rikvdh made a start with this, but has had to step back.

**Comment by zigi05** [comment]
Guys, any update on this?

**Comment by JasonTheAdams** [comment]
Hoping to see progress on this as well

**Comment by screamer-gi** [comment]
I want to participate in the project. Without supporting PER CS 2.0 / PER CS 3.0 by PHPCS I can't persuade my teammates to switch to PER CS from PSR-12

**Comment by jrfnl** [comment]
@screamer-gi Thank you for your offer to help with this.

I can give you access to the (for now, private) repo and if you like, we can set up a call to talk things through and get you started ?
Please DM me on [Mastodon](https://phpc.social/@jrf_nl) or [X](https://x.com/jrf_nl) if you want to schedule something.

**Comment by screamer-westland** [comment]
I've messaged you at X. Mastodon didn't approve my registration. You can
reach me at www.linkedin.com/in/george-istomin-7a010420 if no alternative

чт, 24 июл. 2025 г. в 00:14, Juliette ***@***.***>:

**Comment by stlrnz** [comment]
May I ask if there is any progress in creating a PER 3.0 ruleset?
I found https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/1265 and hope it was not discarded but moved into the private repository?

**Comment by screamer-gi** [comment]
Yes, it was moved to the private repo, I'm working on it, creating unit tests first to cover all PER CS rules. Two sections left at this moment

**Comment by JohnnyWalkerDigital** [comment]
Hello! Is there any PER support yet? Feels like just having one evolving standard rather than writing three separate ones would better support real life situations as each iteration of PER simply addresses new PHP features?

Also, I see [Laravel Pint/PHP-CS-Fixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer) and [Mago](https://github.com/carthage-software/mago) both support PER as one standard?

**Comment by JohnnyWalkerDigital** [comment]
> Yes, it was moved to the private repo, I'm working on it, creating unit tests first to cover all PER CS rules. Two sections left at this moment

Why not place it on a branch so we can all contribute?

**Comment by jrfnl** [comment]
> Why not place it on a branch so we can all contribute?

As per the discussion above, the PER standards will not live in this repository, but will be maintained in a separate repo.

I could make that repo public already, but an empty repo is not a good starting point for community contributions. In my opinion, the initial framework (= ruleset + tests + issues about what's missing) needs to be set up first, before the repo can be opened up for drive-by contributions. And unfortunately, that's just not there yet.

**Comment by JohnnyWalkerDigital** [comment]
Thanks for your reply. FWIW I don't think hosting the PER rules in a separate repo makes much sense. 

PER is an evolving standard, yes, but looking at its historical progression, it is not adding lots of new rules on a regular basis:

* PER 1.0: June 2022 (actually just PSR-12)
* PER 2.0: Nov 2023  - [Migration guide from 1.0 to 2.0](https://www.php-fig.org/per/coding-style/meta/migration-3.0/): 11 new rules
* PER 3.0: July 2025 - [Migration guide from 2.0 to 3.0](https://www.php-fig.org/per/coding-style/meta/migration-3.0/): 10 new rules

As you can see, in the past decade (since the release of PSR-12 in January 2016) there has only been 21 new rules added, and many of these rules exist as options already.

What am I missing?

**Comment by jrfnl** [comment]
@JohnnyWalkerDigital Already asked and answered in the discussion above. See: https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/29#issuecomment-2223647593

**Comment by JohnnyWalkerDigital** [comment]
PHP Code Sniffer remains the best and most comprehensive PHP linter out there, thanks to the hard work of the contributors, especially yourself @jrfnl, but without PER support I worry it's being left behind.

My concern is that the current approach of a separate private repo is not proving as fruitful as it could be.

Perhaps I could be added to the private repo to see if I can assist?

**Comment by jrfnl** [comment]
@JohnnyWalkerDigital Happy to add you and I agree that there has been too little progress. After your query last week, I was thinking of having a look myself again later this week. If you'd be interested and would have the time, we could maybe have a look together in a video call ?

**Comment by JohnnyWalkerDigital** [comment]
@jrfnl Yes, it might spark some momentum if we jumped on a video call 👍

**Comment by jrfnl** [comment]
@JohnnyWalkerDigital Are you on Mastodon or X ? If so, please DM me (I'm jrf_nl on both) so we can set that up. Thanks. Looking forward to it.
